container/docker/factory.go

```diff
@@ -197,6 +208,80 @@ func startThinPoolWatcher(dockerInfo *dockertypes.Info) (*devicemapper.ThinPoolW
 	return thinPoolWatcher, nil
 }
 
+func getKernelVersion() (string, error) {
+	unameOutput := syscall.Utsname{}
+	err := syscall.Uname(&unameOutput)
+	if err != nil {
+		return "", err
+	}
+
+	b := make([]byte, len(unameOutput.Release))
+	for i, v := range unameOutput.Release {
+		b[i] = byte(v)
+	}
+	return string(b), nil
+}
+
+func ensureThinLsKernelVersion(kernelVersion string) error {
+	// kernel 4.4.0 has the proper bug fixes to allow thin_ls to work without corrupting the thin pool
+	version_4_4_0 := semver.MustParse("4.4.0")
+	// RHEL kernel 3.10.0 release >= 366 has the proper bug fixes backported from 4.4.0 to allow
+	// thin_ls to work without corrupting the thin pool
+	version_3_10_0 := semver.MustParse("3.10.0")
+
+	matches := version_re.FindStringSubmatch(kernelVersion)
+	if len(matches) < 4 {
+		return fmt.Errorf("error parsing kernel version %q: expected 4 matches, got %d", kernelVersion, len(matches))
+	}
+
+	sem, err := semver.Make(matches[0])
+	if err != nil {
+		return err
+	}
+
+	if sem.GTE(version_4_4_0) {
+		// kernel 4.4+ - good
+		return nil
+	}
+
+	// Certain RHEL/Centos 7.x kernels have a backport to fix the corruption bug
+	if !strings.Contains(kernelVersion, ".el7.") {
+		// not a RHEL 7.x kernel - won't work
+		return fmt.Errorf("kernel version 4.4.0 or later is required to use thin_ls - you have %q", kernelVersion)
+	}
+
+	// RHEL/Centos 7.x from here on
+	if sem.Major != 3 {
+		// only 3.x kernels *may* work correctly
+		return fmt.Errorf("RHEL/Centos 7.x kernel version 3.10.0-366 or later is required to use thin_ls - you have %q", kernelVersion)
+	}
+
+	if sem.GT(version_3_10_0) {
+		// 3.10.1+ - good
+		return nil
+	}
+
+	if sem.EQ(version_3_10_0) {
+		// need to check release
+		releaseRE := regexp.MustCompile(`^[^-]+-([^.]+)\.`)
+		releaseMatches := releaseRE.FindStringSubmatch(kernelVersion)
+		if len(releaseMatches) != 2 {
+			return fmt.Errorf("unable to determine RHEL/Centos 7.x kernel release from %q", kernelVersion)
+		}
```

Comment:
Yes, except for the aforementioned fact that having x86_64 in the version string won't parse as a semver. so my `sem` literally only has the major, minor, and patch values.